### PR TITLE
Update pattern fusing algo for quantization

### PR DESCRIPTION
**Pull request by aleksei-kashapov**
This PR fixes the pattern fusing algorithm. Before, the algorithm tried to find a pattern with the maximum length. It leads to a bug at the next steps where there was a check whether the subgraph has an outgoing edge and, if it has, the algorithm stops processing it, not trying to match smaller patterns.

Now, the algorithm doesn't stop if the maximum pattern doesn't fit and tries to match smaller ones

### CODE REVIEW

**Comment by aleksei-kashapov** [comment]
Jenkins please retry a build

**Review by alexsu52 — COMMENTED** [review]
Looks good to me.

File: nncf/torch/graph/graph_matching.py [hunk 1]
```diff
@@ -44,20 +44,40 @@ def __and__(self, other):
...
+
+                    all_matches.append((n, following))
+                # pylint: enable=E1133
+            else:
+                n, following = result
+                following = list(following)
+                if not isinstance(n, list):
+                    n = [n]
+
+                all_matches.append((n, following))
```
Comment (andrey-churkin) [inline nncf/torch/graph/graph_matching.py:70]:
```suggestion
            if not isinstance(result, list):
                result = [result]
            # https://github.com/PyCQA/pylint/issues/3105
            # pylint: disable=E1133
            for res in result:
                n, following = res
                following = list(following)
                if not isinstance(n, list):
                    n = [n]

                all_matches.append((n, following))
            # pylint: enable=E1133
```

File: nncf/torch/graph/graph_matching.py [hunk 2]
```diff
@@ -44,20 +44,40 @@ def __and__(self, other):
...
+        if not all_matches:
+            return None
```
Comment (andrey-churkin) [inline nncf/torch/graph/graph_matching.py:72]: @aleksei-kashapov  Why we can not return an empty list?

File: nncf/torch/graph/graph_matching.py (hunk 1 above)
Comment (aleksei-kashapov) [reply nncf/torch/graph/graph_matching.py:70]: good idea

File: nncf/torch/graph/graph_matching.py (hunk 2 above)
Comment (aleksei-kashapov) [reply nncf/torch/graph/graph_matching.py:72]: indeed, we can

File: nncf/torch/graph/graph_matching.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/torch/graph/graph_matching.py:70]: I think this might help with the pylint issue.

File: nncf/torch/graph/graph_matching.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/torch/graph/graph_matching.py:72]: Let's do it. Please update your PR.

File: nncf/torch/graph/graph_matching.py [hunk 3]
```diff
@@ -44,19 +44,30 @@ def __and__(self, other):
...
+                result = [result]
+            for res in result:
+                n, following = res
+                following = list(following)
+                if not isinstance(n, list):
+                    n = [n]
+
+                all_matches.append((n, following))
+        if not all_matches:
+            return []
```
Comment (andrey-churkin) [inline nncf/torch/graph/graph_matching.py:63]: it's not necessary
```suggestion
```

File: nncf/torch/graph/graph_matching.py [hunk 4]
```diff
@@ -215,21 +226,76 @@ def get_edge_boundaries(match: List[str], graph: nx.DiGraph):
...
+    nodes = nx.topological_sort(graph)
+    for node in nodes:
+        # If a node has already been added to any pattern skip this node
+        if node in subgraphs_nodes:
+            continue
+
+        all_matches = expression.all_matches([node], graph)
+
+        if all_matches is None:
+            continue
```
Comment (andrey-churkin) [inline nncf/torch/graph/graph_matching.py:283]:
```suggestion
```

Comment (aleksei-kashapov) [reply nncf/torch/graph/graph_matching.py]: Done

File: nncf/torch/graph/graph_matching.py (hunk 3 above)
Comment (aleksei-kashapov) [reply nncf/torch/graph/graph_matching.py]: Done

File: nncf/torch/graph/graph_matching.py (hunk 2 above)
Comment (aleksei-kashapov) [reply nncf/torch/graph/graph_matching.py]: Done

File: nncf/torch/graph/graph_matching.py (hunk 1 above)
Comment (aleksei-kashapov) [reply nncf/torch/graph/graph_matching.py]: Done

**Review by vshampor — APPROVED** [review]

**Comment by aleksei-kashapov** [comment]
Jenkins please retry a build

**Review by andrey-churkin — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]

**Pull request merged**